lms/services/assignment.py

```diff
@@ -10,55 +15,206 @@ def __init__(self, db):
+
+    def _get_by_resource_link_id(self, tool_consumer_instance_guid, resource_link_id):
+        return (
+            self._db.query(Assignment)
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                resource_link_id=resource_link_id,
+            )
+            .one()
+        )
+
+    def _get_by_ext_lti_assignment(
+        self, tool_consumer_instance_guid, ext_lti_assignment_id
+    ):
+        return (
+            self._db.query(Assignment)
+            .filter_by(
+                tool_consumer_instance_guid=tool_consumer_instance_guid,
+                ext_lti_assignment_id=ext_lti_assignment_id,
+            )
+            .one()
+        )
+
+    def _get_for_canvas_launch(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id,
+        ext_lti_assignment_id,
+    ):
+        """
+        Get a canvas assignment by both resource_link_id and ext_lti_assignment_id.
+
+        Having both values means that we are most likely launching an assignment but
+        we might find these cases on the DB data:
+
+            - A row with both values.
+                An assignment that has been launched before.
+                Just get it an return it.
+
+            - A row with with just the `ext_lti_assignment_id`.
+                We have configured this assignment before but never launched it
+                so we don't yet know its `resource_link_id`.
+
+                Get it, fill the new value on `resource_link_id` and return it.
+
+            - Two rows, one with each of the ids an the other set to null.
+
+                In this case we stored the assignment in the past on launch with just the `resource_link_id`,
+                later, we re-configured it, creating a new row with just the `ext_lti_assignment_id`.
+
+                We need to get both and merge them into one.
+                We'll keep the one with `ext_lti_assignment_id` as is the newest one.
+        """
+        query = self._db.query(Assignment).filter(
+            Assignment.tool_consumer_instance_guid == tool_consumer_instance_guid,
+            (
+                # Regular canvas launch
+                (
+                    (Assignment.resource_link_id == resource_link_id)
+                    & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                )
+                # Configured file assignment that was stored in the DB
+                # before all canvas assignments were stored in the DB
+                | (
+                    (Assignment.resource_link_id == resource_link_id)
+                    & (Assignment.ext_lti_assignment_id.is_(None))
+                )
+                #  First launch of a newly configured assignment
+                | (
+                    (Assignment.resource_link_id.is_(None))
+                    & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                )
+            ),
+        )
```

Comment:
The `in` behaves weirdly with null as anything that compares against it (`select 1 where null = null;`) so I think we need to use `is null` (sqlalcehmy converts `x == null` to `x is null` so that will work too) explicitly.